Mobile Buy SDK/Mobile Buy SDK/Operations/BUYCheckoutOperation.m

```diff
@@ -0,0 +1,168 @@
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+#import "BUYCheckoutOperation.h"
+#import "BUYClient+Checkout.h"
+#import "BUYPaymentToken.h"
+#import "BUYCheckout.h"
+#import "BUYRequestOperation.h"
+
+@interface BUYClient (PrivateCheckout)
+
+- (BUYRequestOperation *)beginCheckout:(BUYCheckout *)checkout paymentToken:(id<BUYPaymentToken>)paymentToken completion:(BUYDataCheckoutBlock)block;
+- (BUYRequestOperation *)getCompletionStatusOfCheckoutToken:(NSString *)token start:(BOOL)start completion:(BUYDataStatusBlock)block;
+- (BUYRequestOperation *)getCheckout:(BUYCheckout *)checkout start:(BOOL)start completion:(BUYDataCheckoutBlock)block;
+
+- (void)startOperation:(BUYOperation *)operation;
+
+@end
+
+@interface BUYCheckoutOperation ()
+
+@property (strong, nonatomic, readonly) BUYCheckout *checkout;
+@property (strong, nonatomic, readonly) id<BUYPaymentToken> token;
+@property (strong, nonatomic, readonly) BUYCheckoutOperationCompletion completion;
+
+@property (strong, nonatomic) BUYRequestOperation *currentOperation;
+
+@end
+
+@implementation BUYCheckoutOperation
+
+#pragma mark - Init -
+
++ (instancetype)operationWithClient:(BUYClient *)client checkout:(BUYCheckout *)checkout token:(id<BUYPaymentToken>)token completion:(BUYCheckoutOperationCompletion)completion
+{
+	return [[[self class] alloc] initWithClient:client checkout:checkout token:token completion:completion];
+}
+
+- (instancetype)initWithClient:(BUYClient *)client checkout:(BUYCheckout *)checkout token:(id<BUYPaymentToken>)token completion:(BUYCheckoutOperationCompletion)completion
+{
+	self = [super init];
+	if (self) {
+		_client     = client;
+		_checkout   = checkout;
+		_token      = token;
+		_completion = completion;
+	}
+	return self;
+}
+
+#pragma mark - Finishing -
+
+- (void)finishWithCheckout:(BUYCheckout *)checkout
+{
+	[self finishExecution];
+	self.completion(checkout, nil);
+}
+
+- (void)finishWithError:(NSError *)error
+{
+	[self finishExecution];
+	self.completion(nil, error);
+}
+
+#pragma mark - Execution -
+
+- (void)startExecution
+{
+	if (self.isCancelled) {
+		return;
+	}
+	
+	[super startExecution];
+	
+	__weak typeof(self) weakSelf = self;
+	
+	BUYRequestOperation *beginOperation = [weakSelf.client beginCheckout:weakSelf.checkout paymentToken:weakSelf.token completion:^(BUYCheckout *checkout, NSError *error) {
+		if (weakSelf.isCancelled) {
+			return;
+		}
+		
+		if (!checkout) {
+			[weakSelf finishWithError:error];
+			return;
+		}
+		
+		BUYRequestOperation *pollOperation = [weakSelf.client getCompletionStatusOfCheckoutToken:weakSelf.checkout.token start:NO completion:^(BUYStatus status, NSError *error) {
```

Comment:
The only glue that binds the operations together is the checkout object, which should be strongly held by this class.  

This does bring up another point, however, of the confusion of having the operations require an checkout object as an input, and returning another checkout object in its response.  It would be clearer if only the checkout token was required to instantiate each operation (extract the `free` check elsewhere), which should make it clearer that the response is a new object.

As for this implementation, each of the operations could be enqueued with the checkout token (which would not change) marking each as a dependancy on the next.  The problem is if a dependancy is canceled, the cancellation does not propagate to the dependants.  A GroupOperation class could be implemented which incapsulates the necessary logic for this to work; setting up the dependancies and propagating cancellations.